orderbook/src/database/app_data.rs

```diff
@@ -0,0 +1,225 @@
+use super::*;
+use anyhow::Result;
+use futures::stream::TryStreamExt;
+use futures::StreamExt;
+use model::app_data::AppDataBlob;
+use primitive_types::{H160, H256};
+use std::borrow::Cow;
+use thiserror::Error;
+
+#[async_trait::async_trait]
+pub trait AppDataStoring: Send + Sync {
+    async fn insert_app_data(&self, app_data: &AppDataBlob) -> Result<H256, InsertionError>;
+    fn app_data<'a>(&'a self, filter: &'a AppDataFilter) -> BoxStream<'a, Result<AppDataBlob>>;
+}
+
+/// Any default value means that this field is unfiltered.
+#[derive(Clone, Default)]
+pub struct AppDataFilter {
+    pub app_data_hash: Option<H256>,
+    pub referrer: Option<H160>,
+}
+
+#[derive(Error, Debug)]
+pub enum InsertionError {
+    #[error("Parsing the string was not successful:")]
+    ParsingStringError(#[from] serde_json::Error),
+    #[error("Duplicated record for `{0}`:")]
+    DuplicatedRecord(H256),
```

Comment:
Not an issue, but curious if you thought about just ignoring duplicated records as idempotent?

Reply:
I would think that the user would want to know if there was a duplicated record rather than just swallowing a second request, nothing doing anything and then not telling someone about it. Gives the front end the opportunity to say something like "yes we already know and have recorded your request".

Reply:
The other option would be to return 200 if the value already exist. Does the user care if they have previously stored the same referral data (would that change something about their workflow)? I'd think they want to tell the backend: "fyi here is the information attached to that app_id hash, if you already have it fine, if not store it", but I don't think they need to know which of the two is the case.

But again, this is not a big issue.

Reply:
I would actually suggest erroring with a [403 - Already Exists](https://developers.rebrandly.com/docs/403-already-exists-errors).

Reply:
My plan was to usually return the STATUSCODE::CREATED, but if it already exists, I would return STATUSCODE::OK and not an error, as this is a legit usecase as Felix is elaborating